### PR TITLE
OCEANWATER-908 Power fault fixes

### PR DESCRIPTION
**Pull request by kmdalal**
## Linked Issues:

| Jira Ticket 🎟️ | [OCEANWATER-908](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-908) |
| Github :octocat:  | #230 |

## Summary of Changes
* Synchronized the processing of power fault profiles with the GSAP processing rate. For the current GSAP rate of .5 Hz, this means using every other line of the profile.
* If a fault profile is consumed, issue a warning and keep reusing the last line, instead of restarting from the beginning.
* Removed the simulation of delayed faults in the profiles; faults occur upon injection.  _Note that this gives all three power faults practically identical behavior._ 
* Capped the power input to GSAP at 30W.
* Introduced a baseline power draw of 1W.
* Averages, not accumulates, the wattage to batch process, for arm operation.
* Some related code abstraction, reformatting (shortening lines) and variable-renaming.
* Revised power fault tests 9.2.2 - 9.2.5 , for Release 10: https://babelfish.arc.nasa.gov/confluence/display/OCEANWATERS/OW-TEST-010-9+Fault+Injection+and+Detection

## Test
* Perform tests 9.2.2 - 9.2.5 at the Confluence link given above.
* NOTE: fault profile exhaustion is not tested.  Could add a test if desired.
* NOTE: the lander is idle during the tests.  Have tested with continuous load and observed faster battery death as expected.  Could add this test (and associated plots) if desired.
* NOTE: Remaining useful life (RUL) is not tested yet because it is broken (goes to 0 almost immediately upon fault).  This problem pre-existed in Release 9 and is addressed by OCEANWATER-315.

## Review suggestions
* Code review: Ussama and Thomas
* Test: Anjan, and either Ussama or Thomas

### CODE REVIEW

**Comment by AstroStucky** [comment]
For 2.2.4, testing the instantaneous capacity loss power fault, SoC and battery temp behave as expected, but `/faults/power_faults_status` stayed 0 for the first few minutes of the test. It did eventually change to 1 before the voltage/temperature thresholds were reached.

**Comment by AstroStucky** [comment]
> For 2.2.4, testing the instantaneous capacity loss power fault, SoC and battery temp behave as expected, but `/faults/power_faults_status` stayed 0 for the first few minutes of the test. It did eventually change to 1 before the voltage/temperature thresholds were reached.

This also happens for the thermal fault test.

**Comment by kmdalal** [comment]
> For 2.2.4, testing the instantaneous capacity loss power fault, SoC and battery temp behave as expected, but `/faults/power_faults_status` stayed 0 for the first few minutes of the test. It did eventually change to 1 before the voltage/temperature thresholds were reached.

I have clarified the expected behavior in the release test, which is that flag becomes when a threshold or delta is reached, and not when the fault is injected.  So it sounds like you're seeing the expected behavior.  I'd like to revisit this specification in our next dev day.

File: ow_power_system/src/power_system_node.cpp
Comment (AstroStucky) [inline ow_power_system/src/power_system_node.cpp:193]: Based on the `/joint_states` publish rate of 50Hz, the GSAP prognoser rate of 0.5 Hz, and a moving average window size of 10 datapoints, the prognoser only ever sees a mean mechanical power that's calculated from the 10 most recent datapoints out of the 25 new datapoints available to it each cycle. I am wondering if it would be more preferable to compute the mean for all 25 new datapoints?

The easiest way to achieve this is to change the moving average window size to 25.

File: ow_power_system/include/power_system_node.h
Comment (AstroStucky) [inline ow_power_system/include/power_system_node.h:46]: `powerCb` contains the "Cb" abbreviation, but is never used as a callback function. And I think it could use a more descriptive name.

File: ow_power_system/src/power_system_node.cpp
Comment (AstroStucky) [inline ow_power_system/src/power_system_node.cpp:357]: If the baseline power accounts for always-on subsystems (computer, heater, ect.) shouldn't it be added to the total electrical power?

**Review by AstroStucky — COMMENTED** [review]
@kmdalal clarified that I was seeing expected behavior, so all tests have passed! :confetti_ball: 

Holding off on approval until we've had the chance to discuss some of my code comments.

File: ow_power_system/include/power_system_node.h
Comment (kmdalal) [reply ow_power_system/include/power_system_node.h:46]: Pre-existing code, but good catch.  I've renamed it to runPrognoser().

File: ow_power_system/include/power_system_node.h
Comment (kmdalal) [reply ow_power_system/include/power_system_node.h:79]: fixed

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:20]: Done, though with more descriptive prefix.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:193]: Great catch!  Your reasoning makes sense to me and I've changed the window to 25.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:357]: I was interpreting it as "minimum power" but I think your take is right.  Changed, and included a comment.

**Comment by kmdalal** [comment]
I've addressed all of @AstroStucky 's code comments so far, retested, and observe virtually the same behavior.  I will regenerate the plots and compare with the old ones, and update the test page if the new plots are noticeably different.   But I'll hold off on this until after the code review is finished.

File: ow_power_system/src/power_system_node.cpp
Comment (AstroStucky) [inline ow_power_system/src/power_system_node.cpp:270]: I think whatever causes the prognoser to behave erratically when power goes over 30W should be investigated in another ticket, if we don't have one for it already (Jira won't load for me right now, so I can't check).

I am also wondering if it would make sense to log a warning whenever `MAX_GSAP_INPUT_WATTS` is used instead of the actual mechanical power. Thoughts?

**Review by AstroStucky — CHANGES_REQUESTED** [review]
Retested the first fault test, and it worked fine. I see a potential need for a warning message to be logged, but I'll leave it up to discussion since I'm not very familiar with GSAP.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp]: Agreed on the first comment, but let me just ask Chetan first.

Excellent suggestion in the second comment: implemented, and I tested it by spoofing a high power value in a fault profile.

**Review by AstroStucky — APPROVED** [review]
All my comments were addressed, and I'm happy with the solutions posed. Great work!

File: ow_power_system/src/power_system_node.cpp
Comment (AstroStucky) [reply ow_power_system/src/power_system_node.cpp]: Thanks! I never did bother to test the warning message. Looks good to me.

By the way, I do not know why this comment got duplicated. GitHub seems to exhibit some weird bugs but only ever when I use it on my Linux machine.

**Review by Samahu — CHANGES_REQUESTED** [review]
Recommend to remove **m_fault_profile_exhausted** and replace the ROS_WARN_STREAM with ROS_WARN_STREAM_ONCE

File: ow_power_system/src/power_system_node.cpp
Comment (Samahu) [reply ow_power_system/src/power_system_node.cpp:38]: Per our discussion today this is a temporary hack till we have a better way for handling excessive  power loads

**Review by Samahu — CHANGES_REQUESTED** [review]
Sorry I forgot to publish my review during our meeting .. here are some of my comments:

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:34]: Makes sense.  Can you point me to that file?

File: ow_power_system/include/power_system_node.h
Comment (kmdalal) [reply ow_power_system/include/power_system_node.h:67]: I didn't know about the "ONCE" versions of ROS logging.  Indeed, that's the way to go.  Removed!

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:264]: I think the message is useful in indicating that a profile was actually not long enough, as unlikely as that may be.  Replaced this approach with the ONCE log - thanks again for that tip.

File: ow_power_system/include/power_system_node.h
Comment (kmdalal) [reply ow_power_system/include/power_system_node.h:67]: Done.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:26]: Good observation, though the actual cast is float to double, and casting either way seems harmless in this case.  I changed the type to double so that everything is consistent.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:38]: I've added a detailed comment to this effect.  Don't have new information to change anything yet.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:45]: Done.

File: ow_power_system/src/power_system_node.cpp
Comment (Samahu) [reply ow_power_system/src/power_system_node.cpp]: Yes, the other line `  ros::Rate rate(GSAP_RATE_HZ);` was also casting the value back to **double**. It is not harmful but not proper.

File: ow_power_system/src/power_system_node.cpp
Comment (Samahu) [reply ow_power_system/src/power_system_node.cpp:34]: The file is called `system.cfg` and is found in the config folder of the ow_power_system.

**Comment by kmdalal** [comment]
No one has yet mentioned that the name of the C++ files should be PowerSystemNode.* per our coding guidelines.  I plan to do this rename as a separate, minor change after this branch is approved.  I didn't want to mess up the diffs by renaming the files as well, though perhaps GitHub would have still shown the correct diffs.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp]: I moved this and several other constants into the system config, per our Dev Day discussion.

**Comment by Samahu** [comment]
While testing the latest changes, I had a sudden drop in the estimated Remaining Useful Life as shown below:
image

I merely enabled the low state of charge fault and the RUL dropped suddenly to zero! Is this still the expected behavior?

**Comment by Samahu** [comment]
> While testing the latest changes, I had a sudden drop in the estimated Remaining Useful Life as shown below: image
> 
> I merely enabled the low state of charge fault and the RUL dropped suddenly to zero! Is this still the expected behavior?

Same thing happened when attempting the instantaneous capacity loss

**Comment by kmdalal** [comment]
@Samahu , yes RUL still drops suddenly and is addressed by a different Jira.  And all 3 power faults are now almost the same.  These points are in the Test section of the Description.

**Comment by Samahu** [comment]
I was testing the low_state_of_charge fault but eventually I got the thermal error:

image


Please note the `faults/power_faults_status` does not indicate which fault took place but it should be clear from the image which fault check did trigger the fault status (state_of_charge > 10% while temperature > 50). I thought this problem was resolved when we updated the power fault profiles.

**Review by Samahu — COMMENTED** [review]
Some comments are worth considering

**Comment by kmdalal** [comment]
> Please note the `faults/power_faults_status` does not indicate which fault took place but it should be clear from the image which fault check did trigger the fault status (state_of_charge > 10% while temperature > 50). I thought this problem was resolved when we updated the power fault profiles.

All 3 power faults work the same way: they induce a high power draw.  All 3 faults are essentially the same because the power draw is +24W or +25W.  And all 3 faults cause temperature to rise fast, SOC to fall fast.   And you're right, the particular fault type injected is not conveyed or discernable in the telemetry.  There is a separate Jira to address this (design) issue.

**Comment by Samahu** [comment]
> > Please note the `faults/power_faults_status` does not indicate which fault took place but it should be clear from the image which fault check did trigger the fault status (state_of_charge > 10% while temperature > 50). I thought this problem was resolved when we updated the power fault profiles.
> 
> All 3 power faults work the same way: they induce a high power draw. All 3 faults are essentially the same because the power draw is +24W or +25W. And all 3 faults cause temperature to rise fast, SOC to fall fast. And you're right, the particular fault type injected is not conveyed or discernable in the telemetry. There is a separate Jira to address this (design) issue.

I am not arguing that .. I am just noting that I used `low_state_of_charge` fault profile but I got a **thermal fault** instead.

**Comment by kmdalal** [comment]
> I am just noting that I used low_state_of_charge fault profile but I got a thermal fault instead.

Yes, for any of the faults, _either_ the voltage threshold or thermal threshold might get crossed first -- behavior is stochastic.  I'll document this in the user guide.

File: ow_power_system/config/system.cfg
Comment (kmdalal) [reply ow_power_system/config/system.cfg]: Done.

File: ow_power_system/include/power_system_node.h
Comment (kmdalal) [reply ow_power_system/include/power_system_node.h]: Done.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp]: Good catch, done.

File: ow_power_system/src/power_system_node.cpp
Comment (kmdalal) [reply ow_power_system/src/power_system_node.cpp:359]: Done.

**Review by Samahu — APPROVED** [review]
The fact that power fault profiles may not generate the intended fault is rather confusing but it can be addressed in a separate issue out of this PR.

**Comment by kmdalal** [comment]
> The fact that power fault profiles may not generate the intended fault is rather confusing but it can be addressed in a separate issue out of this PR.

This issue is addressed by OCEANWATER-925, and I've just added an explanation of this behavior in the user guide.

**Pull request merged**
